crates/storage/src/connection/trie.rs

```diff
@@ -163,28 +178,29 @@ mod macros {
                         .context("Creating insert statement")?;
 
                     let mut to_insert = Vec::new();
-                    let mut to_process = vec![Child::Hash(root)];
+                    let mut to_process = vec![NodeRef::Index(update.nodes_added.len() - 1)];
 
                     while let Some(node) = to_process.pop() {
-                        // Only hash variants need to be stored.
+                        // Only index variants need to be stored.
                         //
                         // Leaf nodes never get stored and a node having an
                         // ID indicates it has already been stored as part of a
                         // previous tree - and its children as well.
-                        let Child::Hash(hash) = node else {
+                        let NodeRef::Index(idx) = node else {
                             continue;
                         };
 
-                        let node = nodes.get(&hash).context("New node data is missing")?;
-                        to_insert.push(hash);
+                        let (_, node) =
+                            &update.nodes_added.get(idx).context("Node index missing")?;
+                        to_insert.push(idx);
 
                         match node {
                             Node::Binary { left, right } => {
-                                to_process.push(left.clone());
-                                to_process.push(right.clone());
+                                to_process.push(*left);
+                                to_process.push(*right);
                             }
                             Node::Edge { child, .. } => {
-                                to_process.push(child.clone());
+                                to_process.push(*child);
                             }
```

Comment:
@Mirko-von-Leipzig BTW: why do we need tree-traversal here? I think we're already doing insertion into the `nodes_added` vector children-first, so we could just iterate over that vector in the loop below that's actually inserting the nodes?

Reply:
Two reasons:

1. `nodes_added` might contain nodes that aren't actually in the final trie depending on how this vector gets generated. If you add each node as you perform an update then you would get superflous nodes. If you generate this set only during `commit` via tree traversal then you are fine.
2. And then this was the actual reason I think. Since the original used `HashMap` the order was random. And I think originally I had the indices be FKs to itself i.e. one had to insert in reverse order for the FK children indices to exist. The comment line `// Insert nodes in reverse to ensure children always have an assigned index for the parent to use.` hints at this.

I believe you should be fine to remove this.

Reply:
You should still need the children first mantra since you need the indices to populate this. But if you're happy to guarantee this externally then you can drop it. That invariant must be commented though - would be very easy to break this by accident.

Keeping the tree traversal is fairly cheap and means we can ignore enforcing external invariants.